src/restart_counter.rs

```diff
@@ -0,0 +1,104 @@
+use crate::ProbeId;
+use core::fmt;
+
+/// The persistent sequence number value returned when this
+/// probe is not tracking restarts.
+pub const NO_RESTART_TRACKING_SEQUENCE_NUMBER: u16 = 0;
+
+/// A persistent restart sequence counter
+pub trait RestartCounter {
+    /// Get the next persistent sequence number
+    fn next_sequence_id(&mut self, probe_id: ProbeId) -> u16;
```

Comment:
As in the C header file, I think we should talk more in the documentation about what's expected of this sequence provider. E.G. :
* Should the values returned be monotonically increasing? 
* Is there a prescribed step size? 
* How are overflows/wrap-around expected to be handled?
* Is 0 and acceptable value to return?
    * If it's not, then we may consider making this trait return `NonZeroU16` or `Option<NonZeroU16>` rather than `u16`